client/pkg/transport/transport.go

```diff
@@ -31,7 +33,17 @@ func NewTransport(info TLSInfo, dialtimeoutd time.Duration) (*http.Transport, er
 	}
 
 	t := &http.Transport{
-		Proxy: http.ProxyFromEnvironment,
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			// according to the comment of http.ProxyFromEnvironment: if the
+			// proxy URL is "localhost" (with or without a port number),
+			// then a nil URL and nil error will be returned.
+			// Thus, we need to workaround this by manually setting an
+			// ENV named FORWARD_PROXY and parse the URL (which is a localhost in our case)
+			if httpProxy, exists := os.LookupEnv("FORWARD_PROXY"); exists {
```

Comment:
Don't know if we want to support a new environment variable `FORWARD_PROXY` for proxying.

I know that `http.ProxyFromEnvironment(req)` will ignore `HTTP_PROXY` when it URL targets localhost, but what about just changing the `LookupEnv` to check for `HTTP_PROXY` and return early?


cc @ahrtr

Reply:
The reason I did it this way is that I don't want to interfere with the default Golang behavior, since this code change will be reflected in the normally built binaries and I am not sure if this will break some other parts of the system or existing user assumption/usage. (my main concern is that `HTTP_PROXY` or `HTTPS_PROXY` parsing behavior will change if we use `LookupEnv` to intercept the `HTTP_PROXY` or `HTTPS_PROXY` and if it's set with localhost or 127.0.0.1)

I am too new to judge so I need @ahrtr 's help to confirm this!

Reply:
I suggest that we do not change this very basic common function. Instead, we can extend its functionality on top of it in **test** only using the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern).

Specifically, you can add a similar function `NewTransport` in test framework, it calls this function, but it also lookup the environment variable "`HTTP_PROXY`" or "`HTTPS_PROXY`", and set the proxy no matter it's localhost or not.

Reply:
The other option is to enhance the test framework to enable contributors to use a non localhost IP, so that you can use environment variable "HTTP_PROXY" or "HTTPS_PROXY" to set a proxy directly, and no need to change this common function `NewTransport`, similar to what https://github.com/etcd-io/etcd/pull/17661 does.

Reply:
Clarify on my previous comment https://github.com/etcd-io/etcd/pull/17938#discussion_r1595329790, 

We can enable users to inject a customized `NewTransport`, similar to 
https://github.com/etcd-io/etcd/blob/e1244f19d66aff5714913efe2ddfc6a971a8f784/client/pkg/srv/srv.go#L29-L30

Reply:
Thanks @ahrtr for the code snippets! 

After looking into the suggestions, I think it's hard to obtain a non-localhost IP and use it as a proxy during e2e testing, so I went on to look into injecting a customized `NewTransport`.

My understanding is that the e2e test is not like a unit test where we can change the function pointers during the execution of the unit test dynamically (like https://github.com/etcd-io/etcd/pull/17938#discussion_r1595352530) because they are in the same module. Instead, the client code is compiled before the e2e test is run. Therefore, some sort of communication channel is required for the e2e test to indicate its need to change the code path during testing (not as easy as swapping `lookupSRV`), so I use the ENV `E2E_TEST_FORWARD_PROXY_IP` as a way to pass in the proxy address. 

My current attempt is https://github.com/etcd-io/etcd/pull/17938/commits/c93e3fc05d8d23f210a6b624514eeae5e95a7042. The idea is kind of borrowed from https://github.com/etcd-io/etcd/pull/17938#discussion_r1595352530, where the HTTP proxy parsing function in use will be determined during the first `Transport` initialization. I currently can't come up with a cleaner way without having to use a custom ENV though... :(  I am not using `HTTP_PROXY` or `HTTPS_PROXY` as we would like to not affect the default behavior of `http.ProxyFromEnvironment` (not parse `localhost` addresses), and only use a custom `NewTransport` during e2e testing only when `E2E_TEST_FORWARD_PROXY_IP` is set. 

Not sure if this makes sense to you! Happy to know if there is a better way of implementing this. :)